cmd/vfkit/main.go

```diff
@@ -264,3 +276,94 @@ func startIgnitionProvisionerServer(ignitionReader io.Reader, ignitionSocketPath
 	log.Debugf("ignition socket: %s", ignitionSocketPath)
 	return srv.Serve(listener)
 }
+
+// it generates a cloud init image by taking the files passed by the user
+// as cloud-init expects files with a specific name (e.g user-data, meta-data) we check the filenames to retrieve the correct info
+// if some file is not passed by the user, an empty file will be copied to the cloud-init ISO to
+// guarantee it to work (user-data and meta-data files are mandatory and both must exists, even if they are empty)
+// if both files are missing it returns an error
+func generateCloudInitImage(files []string) (string, error) {
+	if len(files) == 0 {
+		return "", nil
+	}
+
+	var userDataReader, metaDataReader io.Reader
+	for _, path := range files {
+		if path == "" {
+			continue
+		}
+		file, err := os.Open(path)
+		if err != nil {
+			return "", err
+		}
+		defer file.Close()
+
+		// based on the file name we save its content on the corresponding var
+		filename := filepath.Base(path)
+		switch filename {
+		case "user-data":
+			userDataReader = file
+		case "meta-data":
+			metaDataReader = file
+		}
+	}
+
+	if userDataReader == nil && metaDataReader == nil {
+		return "", fmt.Errorf("cloud-init needs user-data and meta-data files to work")
+	}
+
+	return createCloudInitISO(metaDataReader, userDataReader)
+}
+
+// It generates a temp ISO file containing the metaData and userData passed by the user
+// It also register an exit handler to delete the file when vfkit exits
+func createCloudInitISO(metaDataReader, userDataReader io.Reader) (string, error) {
+	writer, err := iso9660.NewWriter()
+	if err != nil {
+		return "", fmt.Errorf("failed to create writer: %w", err)
+	}
+
+	defer func() {
+		if err := writer.Cleanup(); err != nil {
+			log.Error(err)
+		}
+	}()
+
+	err = writer.AddFile(userDataReader, "user-data")
+	if err != nil {
+		return "", fmt.Errorf("failed to add user-data file: %w", err)
+	}
+
+	err = writer.AddFile(metaDataReader, "meta-data")
+	if err != nil {
+		return "", fmt.Errorf("failed to add meta-data file: %w", err)
+	}
+
+	isoFile, err := os.CreateTemp("", "vfkit-cloudinit-")
+	if err != nil {
+		return "", fmt.Errorf("unable to create temporary cloud-init ISO file: %w", err)
+	}
+
+	// register handler to remove isoFile when exiting
+	util.RegisterExitHandler(func() {
+		os.Remove(isoFile.Name())
+	})
+
+	outputFile, err := os.OpenFile(isoFile.Name(), os.O_WRONLY|os.O_TRUNC|os.O_CREATE, 0644)
```

Comment:
I'm confused by this code, `os.CreateTemp` returns an `os.File` which we can use to write to the file, do we really need to reopen this file?